### PR TITLE
[MIG] [11.0] web_widget_float_formula

### PR DESCRIPTION
**Pull request by hgminh95**
Hi @pedrobaeza, I have finished the migration with this module. Please review and tell me if you see anything wrong.

### CODE REVIEW

**Comment by pedrobaeza** [comment]
Thanks for the submission, Hoang,

Several things you have to take into account:

* Sign the CLA
* Follow migration guide: https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-11.0. Start from scratch rescuing the commits from other authors, and then copy your changes for making it work on v11.
* I will try to review, but I'm not the only one in OCA for that task :wink: 
* Check Travis status, which is red.

**Comment by hgminh95** [comment]
Thanks for your suggestion,

About the error in Travis build, it seems that the origin is from module `web_responsive`. In the test file `web_responsive/static/tests/js/web_responsive.js`, `odoo.define_section()` is used but apparently it have been removed in Odoo 11. I suggest that we should migrate these tests to the new API.

About the CLA, I will sign and submit tomorrow.

**Comment by pedrobaeza** [comment]
Please rebase as now web_responsive is fixed

**Comment by hgminh95** [comment]
Are you sure about that? I still see that function in the latest commit of `11.0`.

```js
https://github.com/OCA/web/blob/11.0/web_responsive/static/tests/js/web_responsive.js

/* Copyright 2016 LasLabs Inc.
 * License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl). */

odoo.define_section('web_responsive', ['web_responsive'], function(test) {
    "use strict";
...
```

**Comment by pedrobaeza** [comment]
Well, the important thing is that now main branch is green, so rebasing on it, if this is still red, it's because of something included in this PR.

**Comment by hgminh95** [comment]
Not sure if I miss something but currently branch `11.0` is red (https://travis-ci.org/OCA/web/branches). The latest pull request 742 is passed but it is on branch `10.0`.

It seems that the image url in travis ci is not correct. The default branch is `11.0` but the image url is `https://travis-ci.org/OCA/web.svg?branch=10.0`.

UPDATE: I check again and see that main branch only failed `travis_transifex` test. The reason it passed other tests because it doesn't run any test at all. You can check the folder `web_responsive/tests/`, there is no `__init__.py` there so Odoo cannot detect any test case. However the js test file `static/tests/js/web_responsive.js` is still included in `qunit_suite` and it is the reason that my pull request failed.

**Comment by pedrobaeza** [comment]
The problem with main branch is in Transifex stage (which I have removed any way), but nothing with the rest). Are you sure you have rebased?

File: web_widget_float_formula/views/web_widget_float_formula.xml [hunk 1]
```diff
@@ -8,7 +8,7 @@
 
 <odoo>
    <template id="assets_backend" name="web_widget_float_formula Assets" inherit_id="web.assets_backend">
-        <xpath expr="." position="inside">
+        <xpath expr="//t[@t-name='web.assets_backend']" position="inside">
```
Comment (yajo) [inline web_widget_float_formula/views/web_widget_float_formula.xml]: This should be unchanged

**Comment by yajo** [comment]
The problem is #759

**Review by yajo — CHANGES_REQUESTED** [review]

**Comment by hgminh95** [comment]
Sorry for long delay, I have update xml file as @Yajo suggested.

**Review by yung-wang — APPROVED** [review]
I tested the PR locally and it works fine. Please do the guideline stuff so this module can be merged. Thanks

File: web_widget_float_formula/views/web_widget_float_formula.xml (hunk 1 above)
Comment (hgminh95) [reply web_widget_float_formula/views/web_widget_float_formula.xml]: This has been resolved. @Yajo: Could you take a look again?

Thanks.

File: web_widget_float_formula/static/src/js/web_widget_float_formula.js [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+            if (typeof value != 'undefined') {
+                return value;
+            }
+            return false;
+        },
+
+        _compute_result: function() {
+            this._clean_formula_text();
+
+            var input = this.$input.val();
```
Comment (astirpe) [inline web_widget_float_formula/static/src/js/web_widget_float_formula.js:79]: With float fields that are defined as computed fields I get this error while creating a record:

```
Error:
TypeError: this.$input is undefined

http://127.0.0.1:8069/web_widget_float_formula/static/src/js/web_widget_float_formula.js:79
Traceback:
_compute_result@http://127.0.0.1:8069/web_widget_float_formula/static/src/js/web_widget_float_formula.js:79:17
commitChanges@http://127.0.0.1:8069/web_widget_float_formula/static/src/js/web_widget_float_formula.js:31:13
commitChanges/defs<@http://127.0.0.1:8069/web/static/src/js/views/basic/basic_renderer.js:67:20
_.collect@http://127.0.0.1:8069/web/static/lib/underscore/underscore.js:164:24
commitChanges@http://127.0.0.1:8069/web/static/src/js/views/basic/basic_renderer.js:66:20
exec/<@http://127.0.0.1:8069/web/static/src/js/core/concurrency.js:216:31
then/</</<@http://127.0.0.1:8069/web/static/lib/jquery/jquery.js:3276:86
fire@http://127.0.0.1:8069/web/static/lib/jquery/jquery.js:3119:38
add@http://127.0.0.1:8069/web/static/lib/jquery/jquery.js:3165:49
then/</<@http://127.0.0.1:8069/web/static/lib/jquery/jquery.js:3275:77
each@http://127.0.0.1:8069/web/static/lib/jquery/jquery.js:383:49
then/<@http://127.0.0.1:8069/web/static/lib/jquery/jquery.js:3272:49
Deferred@http://127.0.0.1:8069/web/static/lib/jquery/jquery.js:3332:25
then@http://127.0.0.1:8069/web/static/lib/jquery/jquery.js:3271:48
exec@http://127.0.0.1:8069/web/static/src/js/core/concurrency.js:215:20
saveRecord@http://127.0.0.1:8069/web/static/src/js/views/basic/basic_controller.js:167:16
saveRecord@http://127.0.0.1:8069/web/static/src/js/views/form/form_controller.js:191:16
OdooClass.extend/</prototype[name]</<@http://127.0.0.1:8069/web/static/src/js/core/class.js:90:35
_onSave@http://127.0.0.1:8069/web/static/src/js/views/form/form_controller.js:547:9
dispatch@http://127.0.0.1:8069/web/static/lib/jquery/jquery.js:4640:50
add/elemData.handle@http://127.0.0.1:8069/web/static/lib/jquery/jquery.js:4309:41
```

That error can be simply avoided by checking that `this.$input` is defined.

It seems to work fine if you write this block of code this way:

```
            if (typeof this.$input != 'undefined') {
                var input = this.$input.val();

                var formula = this._process_formula(input);
                if (formula !== false) {
                    var value = this._eval_formula(formula);
                    if (value !== false) {
                        this._formula_text = "=" + formula;

                        var decimal_point = core._t.database.parameters.decimal_point;

                        // _setValue
                        this._setValue(value.toString().replace(decimal_point, '.'))
                        this._render();
                    }
                }
            }
```

Could you apply this patch?

**Review by astirpe — CHANGES_REQUESTED** [review]

File: web_widget_float_formula/__manifest__.py [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+# Copyright 2016 LasLabs Inc.
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+{
+    'name': 'Web Widget - Formulas in Float Fields',
+    'summary': 'Allow use of simple formulas in float fields',
+    'version': '11.0.1.0.0',
+    'category': 'Web',
+    'author': 'GRAP, LasLabs, Odoo Community Association (OCA)',
+    'website': 'http://www.grap.coop',
```
Comment (nikul-serpentcs) [inline web_widget_float_formula/__manifest__.py:12]: @hgminh95 URL must be [this](https://github.com/OCA/maintainer-tools/blob/master/template/module/__manifest__.py#L11)

**Review by nikul-serpentcs — CHANGES_REQUESTED** [review]
Minor Change: REDME.rst file generate using https://github.com/OCA/maintainer-tools/tree/master/template/module/readme

File: web_widget_float_formula/README.rst [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+Installation
+============
+
+To install this module, simply follow the standard install process.
+
+Configuration
+=============
+
+No configuration is needed or possible.
```
Comment (yajo) [inline web_widget_float_formula/README.rst:37]: Remove the configuration section if it has nothing relevant.

File: web_widget_float_formula/README.rst [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+  fields.float)
+* Uses the eval() JS function to evaluate the formula
+* Does not do any rounding (this is handled elsewhere)
+* Avoids code injection by applying strict regex to formula prior to eval() 
+  (e.g. "=alert('security')" would not get evaluated)
+
+Installation
+============
+
+To install this module, simply follow the standard install process.
```
Comment (yajo) [inline web_widget_float_formula/README.rst:32]: Remove the installation section if it has nothing relevant.

In fact, you could use the brand new readme generator.

File: web_widget_float_formula/README.rst [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+
+Install and enjoy. A short demo video can be found at 
+http://www.youtube.com/watch?v=jQGdD34WYrA.
+
+.. image:: https://odoo-community.org/website/image/ir.attachment/5784_f2813bd/datas
+   :alt: Try me on Runbot
+   :target: https://runbot.odoo-community.org/runbot/162/10.0
+
+Known Issues / Roadmap
+======================
```
Comment (yajo) [inline web_widget_float_formula/README.rst:50]: Remove this empty section too.

**Review by yajo — APPROVED** [review]
OK, apart from what others have already said.

**Comment by yung-wang** [comment]
I got a strange problem with the calculation if you use the decimal and thousand separator. I'm using the dutch language nl_NL and set the decimal to **,**  and the thousand to **.**   The format of the amount i set to [3,0].
When i try to calculate =300/1,21 I get a very large number. Happens in invoices for example.

**Comment by tarteo** [comment]
@hgminh95 You plan on finishing this?

**Comment by hgminh95** [comment]
Sorry for late reply. I haven't got time to finish yet, as I am no longer following Odoo. It would be great if someone can help with this issue.

**Review by legalsylvain — CHANGES_REQUESTED** [review]
Tested on runbot. doesn't work.

step to reproduce : 
- log into runbot
- install sale module
- try to create a product, and in the sale price, tip "=42+14"

-> computation is not done.

**Updated Comment**
the computation is not done on Monetary field. Seems that work correctly on float field. That's a regression, regarding 8.0 module.

**Comment by bennals** [comment]
Apologies if this is not the right place to ask, but I would be interested in financing the completion of this migration if any OCA developers have the time and inclination.  Please advise how to proceed (or the right place to ask it).

**Comment by alexey-pelykh** [comment]
@hgminh95 do you plan to address this migration further?

**Pull request closed without merging**